// @tapico/node-openfeature-posthog/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  FeatureFlagAttributes: () => FeatureFlagAttributes,
  FeatureFlagMetricAttributes: () => FeatureFlagMetricAttributes,
  LoggingHook: () => LoggingHook,
  NAME: () => NAME,
  OpenTelemetryHook: () => OpenTelemetryHook,
  PostHogProvider: () => PostHogProvider,
  VERSION: () => VERSION
});
module.exports = __toCommonJS(src_exports);

// src/provider.ts
var import_server_sdk = require("@openfeature/server-sdk");
var import_posthog_node = require("posthog-node");

// src/types.ts
var FlagError = {
  MISSING_DISTINCT_ID: "TARGETING_KEY_MISSING"
};

// src/provider.ts
var PostHogProvider = class {
  constructor(options) {
    this.metadata = {
      name: `posthog-provider`
    };
    var _a, _b, _c, _d, _e, _f;
    if (!options.posthogClient && !options.posthogConfiguration) {
      throw new Error(`The 'posthogClient' or 'posthogConfiguration' should be given`);
    }
    if (options.posthogClient) {
      this.client = options.posthogClient;
      this.evaluateLocally = (_b = (_a = options.posthogConfiguration) == null ? void 0 : _a.evaluateLocally) != null ? _b : false;
      this.debugMode = (_d = (_c = options.posthogConfiguration) == null ? void 0 : _c.debugMode) != null ? _d : false;
    } else if (options.posthogConfiguration) {
      const { apiKey, personalApiKey, evaluateLocally, clientOptions } = options.posthogConfiguration;
      if (!apiKey) {
        throw new Error(`Missing the PostHog 'apiKey' is not given`);
      }
      if (!personalApiKey) {
        throw new Error(`Missing the PostHog 'personalApiKey' is not given`);
      }
      this.evaluateLocally = evaluateLocally != null ? evaluateLocally : false;
      this.debugMode = (_f = (_e = options.posthogConfiguration) == null ? void 0 : _e.debugMode) != null ? _f : false;
      this.client = new import_posthog_node.PostHog(apiKey, this.createPosthogClientConfig(personalApiKey, clientOptions));
    } else {
      throw new Error(`Failed to intialise PostHog OpenFeature provider`);
    }
    if (this.client && this.debugMode) {
      this.client.debug(true);
    }
    this.setupEventListeners();
  }
  /**
   * @private
   * Register event listerns to the PostHog client to manage the internal state
   * of the provider
   */
  setupEventListeners() {
  }
  createPosthogClientConfig(personalApiKey, config) {
    const posthogConfig = {
      personalApiKey,
      ...config
    };
    return posthogConfig;
  }
  /**
   * Determines the boolean variation of a feature flag for a context, along with information about
   * how it was calculated.
   *
   * If the flag does not evaluate to a boolean value, then the defaultValue will be returned.
   *
   * @param flagKey The unique key of the feature flag.
   * @param defaultValue The default value of the flag, to be used if the value is not available
   *   from PostHog.
   * @param context The context requesting the flag. The client will generate an analytics event to
   *   register this context with PostHog if the context does not already exist.
   * @returns A promise which will resolve to a ResolutionDetails.
   */
  async resolveBooleanEvaluation(flagKey, defaultValue, context, logger) {
    const details = await this.evaluate(flagKey, defaultValue, context, logger);
    if (typeof details.value === "boolean") {
      const value = details.value;
      return {
        ...details,
        value
      };
    } else {
      throw new import_server_sdk.TypeMismatchError(this.getFlagTypeErrorMessage(flagKey, details.value, "boolean"));
    }
  }
  /**
   * Determines the string variation of a feature flag for a context, along with information about
   * how it was calculated.
   *
   * If the flag does not evaluate to a string value, then the defaultValue will be returned.
   *
   * @param flagKey The unique key of the feature flag.
   * @param defaultValue The default value of the flag, to be used if the value is not available
   *   from PostHog.
   * @param context The context requesting the flag. The client will generate an analytics event to
   *   register this context with PostHog if the context does not already exist.
   * @returns A promise which will resolve to a ResolutionDetails.
   */
  async resolveStringEvaluation(flagKey, defaultValue, context, logger) {
    const details = await this.evaluate(flagKey, defaultValue, context, logger);
    if (typeof details.value === "string") {
      const value = details.value;
      return {
        ...details,
        value
      };
    } else {
      throw new import_server_sdk.TypeMismatchError(this.getFlagTypeErrorMessage(flagKey, details.value, "string"));
    }
  }
  /**
   * Determines the numeric variation of a feature flag for a context, along with information about
   * how it was calculated.
   *
   * If the flag does not evaluate to a numeric value, then the defaultValue will be returned.
   *
   * @param flagKey The unique key of the feature flag.
   * @param defaultValue The default value of the flag, to be used if the value is not available
   *   from PostHog.
   * @param context The context requesting the flag. The client will generate an analytics event to
   *   register this context with PostHog if the context does not already exist.
   * @returns A promise which will resolve to a ResolutionDetails.
   */
  async resolveNumberEvaluation(flagKey, defaultValue, context, logger) {
    const details = await this.evaluate(flagKey, defaultValue, context, logger);
    if (typeof details.value === "number") {
      const value = details.value;
      return {
        ...details,
        value
      };
    } else {
      throw new import_server_sdk.TypeMismatchError(this.getFlagTypeErrorMessage(flagKey, details.value, "number"));
    }
  }
  /**
   * Determines the object variation of a feature flag for a context, along with information about
   * how it was calculated.
   *
   * @param flagKey The unique key of the feature flag.
   * @param defaultValue The default value of the flag, to be used if the value is not available
   *   from PostHog.
   * @param context The context requesting the flag. The client will generate an analytics event to
   *   register this context with PostHog if the context does not already exist.
   * @returns A promise which will resolve to a ResolutionDetails.
   */
  async resolveObjectEvaluation(flagKey, defaultValue, context, logger) {
    const details = await this.evaluatePayload(flagKey, defaultValue, context, logger);
    if (typeof details.value === "object") {
      return {
        ...details,
        value: details.value,
        reason: import_server_sdk.StandardResolutionReasons.TARGETING_MATCH
      };
    } else {
      throw new import_server_sdk.TypeMismatchError(this.getFlagTypeErrorMessage(flagKey, details.value, "object"));
    }
  }
  /**
   * @internal
   * Retrieves the payload of the feature flag
   */
  async evaluatePayload(flagKey, defaultValue, context, logger) {
    var _a, _b;
    const translatedContext = this.translateContext(context);
    if (!translatedContext.targetingKey) {
      if (typeof translatedContext.distinctId !== "undefined") {
        logger.warn(
          `You are mixing 'targetingKey' and 'distinctId' fields in the evaluation context. Please avoid using 'distinctId'.`
        );
      }
      return {
        value: defaultValue,
        errorCode: import_server_sdk.ErrorCode.GENERAL,
        reason: FlagError.MISSING_DISTINCT_ID
      };
    }
    try {
      const flagContext = {};
      if (translatedContext.groups) {
        flagContext.groups = translatedContext.groups;
      }
      if ((_a = translatedContext.context) == null ? void 0 : _a.personProperties) {
        flagContext.personProperties = translatedContext.context.personProperties;
      }
      if ((_b = translatedContext.context) == null ? void 0 : _b.groupProperties) {
        flagContext.groupProperties = translatedContext.context.groupProperties;
      }
      const flagConfig = {
        onlyEvaluateLocally: this.evaluateLocally,
        sendFeatureFlagEvents: true,
        ...flagContext
      };
      const flagResult = await this.client.getFeatureFlag(flagKey, translatedContext.targetingKey, flagConfig);
      const flagPayloadResult = await this.client.getFeatureFlagPayload(
        flagKey,
        translatedContext.targetingKey,
        void 0,
        flagConfig
      );
      if (typeof flagResult === "undefined") {
        return {
          value: defaultValue,
          reason: import_server_sdk.StandardResolutionReasons.DEFAULT,
          errorCode: import_server_sdk.ErrorCode.FLAG_NOT_FOUND
        };
      }
      return {
        value: flagPayloadResult,
        reason: import_server_sdk.StandardResolutionReasons.TARGETING_MATCH
      };
    } catch (err) {
      return {
        value: defaultValue,
        reason: import_server_sdk.StandardResolutionReasons.ERROR,
        errorCode: import_server_sdk.ErrorCode.GENERAL
      };
    }
  }
  /**
   * @internal
   * Returns the value of the given feature flag
   * @param flagKey the unique name of the feature flag
   * @param PosthogInfo the user identifier
   * @returns ResolutionDetails
   */
  async evaluate(flagKey, defaultValue, context, logger) {
    var _a, _b;
    const translatedContext = this.translateContext(context);
    if (!translatedContext.targetingKey) {
      if (typeof translatedContext.distinctId !== "undefined") {
        logger.warn(
          `You are mixing 'targetingKey' and 'distinctId' fields in the evaluation context. Please avoid using 'distinctId'.`
        );
      }
      return {
        value: defaultValue,
        errorCode: import_server_sdk.ErrorCode.GENERAL,
        reason: FlagError.MISSING_DISTINCT_ID
      };
    }
    try {
      const flagContext = {};
      if (translatedContext.groups) {
        flagContext.groups = translatedContext.groups;
      }
      if ((_a = translatedContext.context) == null ? void 0 : _a.personProperties) {
        flagContext.personProperties = translatedContext.context.personProperties;
      }
      if ((_b = translatedContext.context) == null ? void 0 : _b.groupProperties) {
        flagContext.groupProperties = translatedContext.context.groupProperties;
      }
      const flagConfig = {
        onlyEvaluateLocally: this.evaluateLocally,
        sendFeatureFlagEvents: true,
        ...flagContext
      };
      const flagResult = await this.client.getFeatureFlag(flagKey, translatedContext.targetingKey, flagConfig);
      if (typeof flagResult === "undefined") {
        return {
          value: defaultValue,
          reason: import_server_sdk.StandardResolutionReasons.DEFAULT,
          errorCode: import_server_sdk.ErrorCode.FLAG_NOT_FOUND
        };
      }
      return {
        value: flagResult,
        reason: import_server_sdk.StandardResolutionReasons.TARGETING_MATCH
      };
    } catch (err) {
      return {
        value: defaultValue,
        reason: import_server_sdk.StandardResolutionReasons.ERROR,
        errorCode: import_server_sdk.ErrorCode.GENERAL
      };
    }
  }
  validateSchema(object) {
    if (Array.isArray(object)) {
      return false;
    }
    return Object.values(object).every((x) => {
      if (typeof x === "object") {
        return this.validateSchema(x);
      } else {
        return typeof x === "string";
      }
    });
  }
  /**
   * @protected
   */
  translateContext(evalContext) {
    const { targetingKey, groups, context } = evalContext;
    if (groups) {
      const isGroupsValid = this.validateSchema(groups);
      if (!isGroupsValid) {
        throw new Error(`An invalid evaluation context value was given for 'groups'`);
      }
    }
    if (context && context.personProperties) {
      const isValid = this.validateSchema(context.personProperties);
      if (!isValid) {
        throw new Error(`An invalid evaluation context value was given for 'context.personProperties'`);
      }
    }
    if (context && context.groupProperties) {
      const isValid = this.validateSchema(context.groupProperties);
      if (!isValid) {
        throw new Error(`An invalid evaluation context value was given for 'context.groupProperties'`);
      }
    }
    return {
      targetingKey,
      ...groups ? { groups } : {},
      context: {
        ...(context == null ? void 0 : context.groupProperties) ? { groupProperties: context.groupProperties } : {},
        ...(context == null ? void 0 : context.personProperties) ? { personProperties: context.personProperties } : {}
      }
    };
  }
  /**
   * Generates type mismatch error message
   * @returns string
   */
  getFlagTypeErrorMessage(flagKey, value, expectedType) {
    return `Flag value ${flagKey} had unexpected type ${typeof value}, expected ${expectedType}.`;
  }
  /**
   * @inheritdoc
   */
  async onClose() {
    if (this.client) {
      await this.client.shutdown();
    }
  }
  /**
   * @inheritDoc
   */
  get hooks() {
    return [];
  }
};

// src/hooks/OpenTelemetryHook.ts
var import_server_sdk2 = require("@openfeature/server-sdk");
var import_api = require("@opentelemetry/api");

// src/VERSION.ts
var VERSION = "1.1.4";
var NAME = "@tapico/node-openfeature-posthog";

// src/hooks/OpenTelemetryHook.ts
var import_core = require("@openfeature/core");
var FeatureFlagAttributes = {
  FLAG_KEY: "feature_flag.flag_key",
  TARGETING_KEY: "feature_flag.targeting_key",
  CLIENT_NAME: "feature_flag.client.name",
  CLIENT_VERSION: "feature_flag.client.version",
  PROVIDER_NAME: "feature_flag.provider.name",
  PROVIDER_MANAGEMENT_URL: "feature_flag.provider.management_url",
  VARIANT: "feature_flag.evaluated.variant",
  VALUE: "feature_flag.evaluated.value",
  REASON: "feature_flag.evaluated.reason"
};
var FeatureFlagMetricAttributes = {
  FLAG_KEY: "flag_key",
  REASON: "reason",
  ERROR_CODE: "error_code",
  PROVIDER_NAME: "provider_name"
};
var OpenTelemetryHook = class {
  constructor(name) {
    this.name = "open-telemetry";
    this.spanMap = /* @__PURE__ */ new WeakMap();
    this.tracer = import_api.trace.getTracer(`@tapico/node-feature:${name}`, VERSION);
    this.meter = import_api.metrics.getMeter(`@tapico/node-feature:${name}`, VERSION, {
      schemaUrl: "https://openfeature.org/schemas/v1/schema.json"
    });
    this.metricFlags = this.meter.createCounter(`feature_flag`, {
      description: "Number of times a feature flag is being used",
      valueType: import_api.ValueType.INT
    });
    this.metricFailedFlags = this.meter.createCounter(`feature_flag_failed`, {
      description: "Number of times a feature flag evaluation failed",
      valueType: import_api.ValueType.INT
    });
  }
  before(hookContext) {
    const span = this.tracer.startSpan(`feature flag - ${hookContext.flagValueType}`);
    span.setAttributes({
      [FeatureFlagAttributes.FLAG_KEY]: hookContext.flagKey,
      [FeatureFlagAttributes.TARGETING_KEY]: hookContext.context.targetingKey,
      [FeatureFlagAttributes.CLIENT_NAME]: hookContext.clientMetadata.name,
      [FeatureFlagAttributes.CLIENT_VERSION]: hookContext.clientMetadata.version,
      [FeatureFlagAttributes.PROVIDER_NAME]: hookContext.providerMetadata.name
    });
    this.spanMap.set(hookContext, span);
    return hookContext.context;
  }
  after(hookContext, flagValue) {
    var _a, _b, _c;
    if (flagValue.variant) {
      (_a = this.spanMap.get(hookContext)) == null ? void 0 : _a.setAttribute(FeatureFlagAttributes.VARIANT, flagValue.variant);
    } else {
      (_b = this.spanMap.get(hookContext)) == null ? void 0 : _b.setAttribute(FeatureFlagAttributes.VALUE, JSON.stringify(flagValue.value));
    }
    if (flagValue.errorCode) {
      this.metricFailedFlags.add(1, {
        [FeatureFlagMetricAttributes.FLAG_KEY]: hookContext.flagKey,
        [FeatureFlagMetricAttributes.ERROR_CODE]: flagValue.errorCode,
        [FeatureFlagMetricAttributes.PROVIDER_NAME]: hookContext.providerMetadata.name
      });
    }
    this.metricFlags.add(1, {
      [FeatureFlagMetricAttributes.FLAG_KEY]: hookContext.flagKey,
      [FeatureFlagMetricAttributes.REASON]: (_c = flagValue.reason) != null ? _c : import_server_sdk2.StandardResolutionReasons.UNKNOWN,
      [FeatureFlagMetricAttributes.PROVIDER_NAME]: hookContext.providerMetadata.name
    });
  }
  finally(hookContext) {
    var _a;
    (_a = this.spanMap.get(hookContext)) == null ? void 0 : _a.end();
  }
  error(hookContext, error) {
    var _a;
    (_a = this.spanMap.get(hookContext)) == null ? void 0 : _a.recordException(error);
  }
};

// src/hooks/LoggingHook.ts
var import_os = require("os");
var LoggingHook = class {
  constructor() {
    this.name = "logging";
  }
  before(hookContext) {
    console.log(`Running 'before' logger hook for flag: ${hookContext.flagKey}`);
    console.log(JSON.stringify(hookContext.context, void 0, 2));
  }
  after(hookContext, details) {
    console.log(`Running 'after' logger hook for flag: ${hookContext.flagKey}`);
    console.log(`Evaluation details:${import_os.EOL}${JSON.stringify(details, void 0, 2)}`);
  }
  finally(hookContext) {
    console.log(`Running 'finally' logger hook for flag: ${hookContext.flagKey}`);
  }
  error(hookContext, err) {
    console.log(`Running 'error' logger hook for flag: ${hookContext.flagKey}`);
    console.error(err);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FeatureFlagAttributes,
  FeatureFlagMetricAttributes,
  LoggingHook,
  NAME,
  OpenTelemetryHook,
  PostHogProvider,
  VERSION
});
